kintoclient/tests/test_client.py

```diff
@@ -0,0 +1,541 @@
+from unittest2 import TestCase
+import json
+import mock
+
+from kintoclient import (
+    Bucket, Session, Permissions, Collection, Record,
+    DEFAULT_SERVER_URL, create_session, KintoException, BucketNotFound
+)
+
+
+# XXX Put this function in tests/support.py
+def mock_response(session, data=None, permissions=None, headers=None,
+                  error=False):
+    data = data or {}
+    permissions = permissions or {}
+    headers = headers or {}
+    info = {'data': data, 'permissions': permissions}
+    if error:
+        session.request.side_effect = ValueError
+    else:
+        session.request.return_value = (info, headers)
+
+
+def get_record(id=None, data=None, permissions=None):
+    record = mock.MagicMock()
+    record.id = id or '1234'
+    record.data = data or {'foo': 'bar'}
+    record.permissions = permissions or {'read': ['Niko', 'Mat']}
+    return record
+
+
+class BucketTest(TestCase):
+
+    def setUp(self):
+        self.session = mock.MagicMock()
+        mock_response(self.session)
+
+    def test_put_is_issued_on_creation(self):
+        Bucket('testbucket', session=self.session, create=True)
+        self.session.request.assert_called_with('put', '/buckets/testbucket',
+                                                permissions=None)
+
+    def test_get_is_issued_on_retrieval(self):
+        Bucket('testbucket', session=self.session)
+        self.session.request.assert_called_with('get', '/buckets/testbucket')
+
+    def test_collection_is_not_created_for_personal_bucket(self):
+        Bucket('default', session=self.session, create=True)
+        self.session.request.assert_called_with('get', '/buckets/default')
+
+    def test_permissions_are_retrieved(self):
+        mock_response(self.session, permissions={'read': ['phrawzty', ]})
+        bucket = Bucket('testbucket', session=self.session)
+
+        self.assertIn('phrawzty', bucket.permissions.read)
+
+    def test_groups_can_be_created_from_buckets(self):
+        pass
+
+    def test_collections_throw_on_unexisting_bucket(self):
+        exception = KintoException()
+        exception.response = mock.MagicMock()
+        exception.response.status_code = 403
+        exception.request = mock.sentinel.request
+
+        self.session.request.side_effect = exception
+
+        try:
+            Bucket('test', session=self.session)
+        except BucketNotFound as e:
```

Comment:
Thanks for pointing that out.

That's reassuring to see that we make the same mistake, though, the context manager is actually a context manager and not the exception itself, so you'll need to do something like that:

```python
with self.assertRaises(BucketNotFound) as cm:
    Bucket('test', session=self.session)

e = cm.exception
self.assertEquals(e.response, exception.response)
```